app/lib/import_export.rb

```diff
@@ -153,16 +153,13 @@ def update_assessment_symptom_data(data, patients)
     data[:assessments][:checked].delete(:symptoms)
     data[:assessments][:headers].delete('Symptoms Reported')
 
-    # For large queries (e.g. 100k patients) of the `Symptom.where(...) query below
-    # we saw MySQL give up on queries and throw an exception.
-    # The workaround for this is to collect the symptom ids in a set and find them
-    # using batches of patients.
-    symptom_ids = Set.new
-    patients.pluck(:id).in_groups_of(10_000, false) do |batch|
-      symptom_ids << Symptom.where(condition_id: ReportedCondition.where(assessment_id: Assessment.where(patient_id: batch).pluck(:id)).pluck(:id)).pluck(:id)
-    end
-    symptom_names_and_labels = Symptom.where(id: symptom_ids.to_a.flatten).where.not(label: nil).where.not(name: nil)
-                                      .distinct.order(:label).pluck(:name, :label).transpose
+    symptom_names_and_labels = patients.joins(assessments: { reported_condition: :symptoms })
```

Comment:
so I think the problem with this query (where we do a bunch of joining instead of plucking ids) is that the query is noticeably slower (this is something I noticed when working on the export performance ticket)

Would you be able to do benchmark comparisons on the 500k database between:

```
Patient.all.joins(assessments: { reported_conditio
n: :symptoms }).where('symptoms.name IS NOT NULL AND symptoms.label IS NOT NULL').dist
inct.order(:label).pluck(:name, :label).transpose
```

and

```
Symptom.where(condition_id: ReportedCondition.where(assessment_id: Assess
ment.where(patient_id: Patient.all.pluck(:id)).pluck(:id)).pluc
k(:id)).where.not(name: nil).where.not(label: nil).distinct.order(:label).pluck(:name,
 :label).transpose
```

I know the id plucking solution also leads to the problem that @Bialogs solved with his workaround, but maybe we'll have to come up with something else that's clever.

Reply:
I'm going to test something out now to see if it performs better than the joins

Reply:
I tested out two new methods of getting that data and they seem to scale differently as more patients are queried 
```ruby
##
# Comparison (5,000 Patients):
#       new_set_method:        0.1 i/s
#       old_set_method:        0.0 i/s - 1.12x  (± 0.00) slower
#         joins_method:        0.0 i/s - 5.46x  (± 0.00) slower
##

##
# Comparison (100,000 Patients):
#       new_set_method:        0.0 i/s
#       old_set_method:        (MySQL has gone away!)
#         joins_method:        0.0 i/s - 1.55x  (± 0.00) slower
##

##
# Comparison (500,000 Patients):
#       new_set_method:        0.0 i/s - 3.07x  (± 0.00) slower
#       old_set_method:        (MySQL has gone away!)
#         joins_method:        0.0 i/s
##


require 'benchmark/ips'

module Benchmarking
  @patients = Patient.all.limit(5_000)

  def joins_method
    symptom_names_and_labels = @patients.joins(assessments: { reported_condition: :symptoms })
                                       .where('symptoms.label IS NOT NULL')
                                       .where('symptoms.name IS NOT NULL')
                                       .distinct
                                       .order(:label)
                                       .pluck(:name, :label)
                                       .transpose
  end

  def old_set_method
    symptom_ids = Set.new
    @patients.pluck(:id).in_groups_of(10_000, false) do |batch|
      symptom_ids << Symptom.where(condition_id: ReportedCondition.where(assessment_id: Assessment.where(patient_id: batch).pluck(:id)).pluck(:id)).pluck(:id)
    end
    symptom_names_and_labels = Symptom.where(id: symptom_ids.to_a.flatten).where.not(label: nil).where.not(name: nil)
                                      .distinct.order(:label).pluck(:name, :label).transpose
  end

  def new_set_method
    symptoms = Set.new
    @patients.pluck(:id).in_groups_of(10_000, false) do |batch|
      Symptom.where(
        condition_id: ReportedCondition.where(
          assessment_id: Assessment.where(
            patient_id: batch
          ).pluck(:id)
        ).pluck(:id)
      ).where.not(label: nil).where.not(name: nil).distinct.pluck(:name, :label).each { |symptom| symptoms.add(symptom) }
    end
    symptom_names_and_labels = symptoms.to_a.sort { |a, b| a[1] <=> b[1] }.transpose
  end

  def benchmark
    Benchmark.ips do |x|
      x.report('joins_method') { joins_method }
      # x.report('old_set_method') { old_set_method }
      x.report('new_set_method') { new_set_method }
      x.compare!
    end
  end
end

``

Reply:
hmm, would switching between the new_set_method and joins_method depending on the patient size between exports be a viable option?

Reply:
So the joins method is certainly cleaner, but I think my preference is for the new_set_method. Most exports are going to be in the range where new_set_method wins, and the 3x difference on the super large export is tolerable. The 5.5x on joins observed on the 5,000 patient export is pretty high.

However, I'll also note that there's a chance we would see meaningfully better results for the joins method in prod as the database servers probably have better specs than our dev machines.

Reply:
@tstrass Is there any way we could conduct any testing on how this would perform in production?

Reply:
@jkufro Not readily

Reply:
But I really don't like the 5x on the 5000 patients so the new_set_method seems safer for now

Reply:
It's hard to compare the largest case because I'm not sure what it's 3x. Is it 3x longer than an hour? is it 3x longer than 15 minutes?

Reply:
I'm thinking we can go with the new_set_method for 1.31.0 (just to have something to fix the problem) and continue to think of alternative solutions over the next couple of weeks? I'm hoping we might be able to come up with something more efficient long term that doesn't require such an expensive query to be made just to get the symptom column names

Reply:
<img width="559" alt="Screen Shot 2021-06-03 at 15 36 50" src="https://user-images.githubusercontent.com/3009651/120721097-08f50500-c482-11eb-9604-e6bd5ceea82d.png">

Reply:
FWIW here's what I got using `Benchmark.measure` on 500,000 patients:

joins_method:
```
user       system     total       real
13.907065  2.937389   16.844454   (117.795818)
```
new_set_method
```
user       system     total       real
81.948856  20.190281  102.139137  (576.950755)
```

Reply:
I'm thinking a hybrid approach with both algorithms cutting over around 250k patients in the export.

Something I remembered too - this really doesn't scale based on number of patients. It scales on the number of symptoms which we use patients a proxy for in these measurements. Before we make a decision we should compare the amount of symptoms per assessment in the test data versus what is reasonable in production.

Reply:
@Bialogs Do those numbers for new_set_method concern for prod?

Reply:
The 1m database has an average of 37 symptoms per patient. That might be a little low, but I'm not sure.